lib/cluster-def-gem/lib/cluster_def.rb

```diff
@@ -0,0 +1,152 @@
+      else
+        entry[:hostname]
+      end
+    end
+
+    def get_entry(name)
+      fetch_cluster_def.select do |ee|
+        ee[:hostname] == name || fqdn(ee) == name
+      end.first
+    end
+
+    def validate_node_number?(nn)
+      # node number must either be '-' or a positive integer
+      # 1..255
+      if nn != '-' and nn.to_i < 1 or nn.to_i > 255 then
+        false
+      else
+        true
+      end
+    end
+
+    def validate_cluster_def(cluster_def, fields)
+        cdef_copy = cluster_def.select{ |row| row[:runlist] != 'SKIP' }
+        # validate columns each row has the same number of fields as fields
+        faulty_rows = cdef_copy.select{ |row| row.length != fields.length }
+        if faulty_rows.length > 0  then
+          faulty_rows.each { |row| puts row }
+          fail "Retreived cluster data appears to be invalid -- missing columns"
+        end
+        # validate node ids 
+        if (cluster_def.select{ |row| validate_node_number?(row[:node_id]) == false }).length > 0  then
+          fail "Retreived cluster data appears to be invalid -- node IDs must be positive integers between 0 and 256 (1..255)"
+        end 
+    end
+
+    def parse_cluster_def(cluster_def)
+      # parse something that looks like cluster.txt and memorize the result
+      fields = [
+                :node_id,
+                :hostname,
+                :mac_address,
+                :ip_address,
+                :ilo_address,
+                :cobbler_profile,
+                :dns_domain,
+                :runlist
+               ]
+
+        # This is really gross because Ruby 1.9 lacks Array#to_h.
+        cdef = cluster_def.map do |line|
+          entry = Hash[*fields.zip(line.split(' ')).flatten(1)]
+          entry.merge({fqdn: fqdn(entry)})
+        end
+        # field size check will fail if we do not do this
+        fields += [:fqdn]
+        validate_cluster_def(cdef, fields)
+        cdef
+    end
+
+    # combines local cluster.txt access with http call to cluster data
+    def fetch_cluster_def
+      if @cluster_def != nil then
+        @cluster_def
+      end
+
+      begin
+        @cluster_def = fetch_cluster_def_http
+      rescue Exception => http_e
+        puts http_e 
+        puts http_e.backtrace
+        @cluster_def = fetch_cluster_def_local 
+      end
+
+      @cluster_def
+    end
+
+    # fetch cluster definition via http
+    def fetch_cluster_def_http
+      cluster_def_url = "http://#{@node_obj[:bcpc][:bootstrap][:server]}#{@node_obj[:bcpc][:bootstrap][:cluster_def_path]}"
+      response = Faraday.new(:proxy => '').get cluster_def_url
+      if response.success? then
+        parse_cluster_def(response.body.split("\n"))
+      else
+        nil
+      end
+    end 
+      
+    # locally access cluster.txt
+    def fetch_cluster_def_local
+      parse_cluster_def(File.readlines(File.join(repo_dir, 'cluster.txt')))
+    end
+
+    def repo_dir
+      '/home/vagrant/chef-bcpc'
```

Comment:
Again same comment about hard-coding `vagrant` as above.

Reply:
We currently have vagrant hardcoded all over the place, I suppose I can change this one if you prefer, but there are a lot of places where it is used